make PID watching a bit less expensive

poettering opened the pull request:
A proposed fix for #9138. Would be good if @davide125 and colleagues could give this a whirl to check that this helps a bit.

davide125 commented:
We had this running for about a day on the affected host in #9138, and so far it seems to be helping: the box is behaving normally and we're not seeing the deluge of `kill(_, 0)` in strace anymore. I'll let this sit over the weekend and report back early next week.

keszybz commented on src/core/unit.c:
```diff
@@ -2622,6 +2627,76 @@ void unit_tidy_watch_pids(Unit *u, pid_t except1, pid_t except2) {
+         * involves issuing kill(pid, 0) on all processes we watch. */
+
+        if (!u->rewatch_pids_event_source) {
+                sd_event_source *s = NULL;
```
This needs `_cleanup_`.

poettering replied:
ouch! indeed!

keszybz commented:
Hmm, `bionic-amd64` failed with a timeout in "upstream". Those tests are pretty flaky, but I haven't seen this failure yet. Let's see what happens when it runs again.

davide125 commented:
Update: our test host is still doing fine with this PR applied

poettering commented:
@davide125 thanks for reporting back!

@keszybz thanks for the review!

force pushed a new version now. Only change: the _cleanup_ decorator is now added, fixing the memory leak i introduced, as pointed out by @keszybz. PTAL!
